From 4075cd52b03f62d23755cd211833b120ca5a5c52 Mon Sep 17 00:00:00 2001
From: Dmitry Dementyev <dementyev@google.com>
Date: Thu, 7 May 2026 13:01:26 -0700
Subject: [PATCH 12/49] Properly parse meta_key in AccountsDb.

Database entries are stored with "auth_uid_for_type:" prefix.
Remove it properly in findMetaAuthUid.

Bug: 507575871
Test: manual
Flag: EXEMPT BUGFIX
Change-Id: I3c028c0cfdbdc44a09fffee3542ba8bc782c1435
(cherry picked from commit 6a2d8b6bc382fa5dc845650d3a47176efa0ef554)
---
 .../core/java/com/android/server/accounts/AccountsDb.java  | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/services/core/java/com/android/server/accounts/AccountsDb.java b/services/core/java/com/android/server/accounts/AccountsDb.java
index 95dbc8ea6311..338aee3aafcd 100644
--- a/services/core/java/com/android/server/accounts/AccountsDb.java
+++ b/services/core/java/com/android/server/accounts/AccountsDb.java
@@ -173,7 +173,8 @@ class AccountsDb implements AutoCloseable {
 
     private static final String META_KEY_FOR_AUTHENTICATOR_UID_FOR_TYPE_PREFIX =
             "auth_uid_for_type:";
-    private static final String META_KEY_DELIMITER = ":";
+    private static final int META_KEY_PREFIX_LEN =
+            META_KEY_FOR_AUTHENTICATOR_UID_FOR_TYPE_PREFIX.length();
     private static final String SELECTION_META_BY_AUTHENTICATOR_TYPE = META_KEY + " LIKE ?";
 
     private final DeDatabaseHelper mDeDatabase;
@@ -1094,8 +1095,8 @@ class AccountsDb implements AutoCloseable {
         Map<String, Integer> map = new LinkedHashMap<>();
         try {
             while (metaCursor.moveToNext()) {
-                String type = TextUtils
-                        .split(metaCursor.getString(0), META_KEY_DELIMITER)[1];
+                String typeWithPrefix = metaCursor.getString(0);
+                String type = typeWithPrefix.substring(META_KEY_PREFIX_LEN);
                 String uidStr = metaCursor.getString(1);
                 if (TextUtils.isEmpty(type) || TextUtils.isEmpty(uidStr)) {
                     // Should never happen.
-- 
2.55.0

